### PR TITLE
implemented juniper ftp client

### PR DESCRIPTION
**Pull request by DFSko**

### CODE REVIEW

**Review by DaniyarAskarov — APPROVED** [review]

File: HappyTravel.Komoro.JuniperChannelManager/Infrastructure/Extensions/ServiceCollectionExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+namespace HappyTravel.Komoro.JuniperChannelManager.Infrastructure.Extensions;
+
+public static class ServiceCollectionExtensions
+{
+    public static void ConfigureJuniperServices(this IServiceCollection services)
+    {
+        services.Configure<JuniperFtpOptions>(o =>
+        {
+            // TODO: add real ftp configuration
```
Comment (suhan3z) [inline HappyTravel.Komoro.JuniperChannelManager/Infrastructure/Extensions/ServiceCollectionExtensions.cs:13]: Should it be like there on production?

File: HappyTravel.Komoro.JuniperChannelManager/Infrastructure/Options/JuniperFtpOptions.cs [hunk 2]
```diff
@@ -0,0 +1,9 @@
+namespace HappyTravel.Komoro.JuniperChannelManager.Infrastructure.Options;
+
+public class JuniperFtpOptions
```
Comment (suhan3z) [inline HappyTravel.Komoro.JuniperChannelManager/Infrastructure/Options/JuniperFtpOptions.cs:3]: Lack of comments

File: HappyTravel.Komoro.JuniperChannelManager/Services/Updater/JuniperFtpClient.cs [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+        if (serializer.Deserialize(stream) is not T entity)
+            throw new Exception("Deserialization failed");
+
+        return entity;
+    }
+
+
+    private readonly Dictionary<Type, string> _patterns = new ()
+    {
+        { typeof(JpHotelInvNotifRQ), "JP_HotelInvNotifRQ_[A-Z0-9]+.xml" }
```
Comment (suhan3z) [inline HappyTravel.Komoro.JuniperChannelManager/Services/Updater/JuniperFtpClient.cs:72]: To constants

**Review by suhan3z — CHANGES_REQUESTED** [review]
According to comments

File: HappyTravel.Komoro.JuniperChannelManager/Infrastructure/Options/JuniperFtpOptions.cs (hunk 2 above)
Comment (DFSko) [reply HappyTravel.Komoro.JuniperChannelManager/Infrastructure/Options/JuniperFtpOptions.cs:6]: Fixed

File: HappyTravel.Komoro.JuniperChannelManager/Infrastructure/Extensions/ServiceCollectionExtensions.cs (hunk 1 above)
Comment (DFSko) [reply HappyTravel.Komoro.JuniperChannelManager/Infrastructure/Extensions/ServiceCollectionExtensions.cs]: moved options to vault

File: HappyTravel.Komoro.JuniperChannelManager/Services/Updater/JuniperFtpClient.cs (hunk 3 above)
Comment (DFSko) [reply HappyTravel.Komoro.JuniperChannelManager/Services/Updater/JuniperFtpClient.cs]: Fixed

File: HappyTravel.Komoro.JuniperChannelManager/Services/Updater/JuniperFtpClient.cs [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+
+
+    private async IAsyncEnumerable<T> Get<T>([EnumeratorCancellation] CancellationToken cancellationToken)
+    {
+        var regex = new Regex(_patterns[typeof(T)], RegexOptions.Compiled);
+
+        foreach (var file in await GetFiles(f => regex.IsMatch(f.Name), cancellationToken))
+        {
+            using var stream = new MemoryStream();
+            await _ftpClient.DownloadStream(stream, file.FullName, token: cancellationToken);
```
Comment (suhan3z) [inline HappyTravel.Komoro.JuniperChannelManager/Services/Updater/JuniperFtpClient.cs:44]: If stream somehow stored in internals of _ftpClient, it won't be disposed

File: HappyTravel.Komoro.JuniperChannelManager/Services/Updater/JuniperFtpClient.cs [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+            yield return Deserialize<T>(stream);
+        }
+    }
+    
+    
+    private async Task<IEnumerable<FtpListItem>> GetFiles(Func<FtpListItem, bool> predicate, CancellationToken cancellationToken)
+    {
+        await _ftpClient.Connect(cancellationToken);
```
Comment (suhan3z) [inline HappyTravel.Komoro.JuniperChannelManager/Services/Updater/JuniperFtpClient.cs:52]: Shouldn't we disconnect?

File: HappyTravel.Komoro.JuniperChannelManager/Services/Updater/JuniperFtpClient.cs (hunk 4 above)
Comment (suhan3z) [reply HappyTravel.Komoro.JuniperChannelManager/Services/Updater/JuniperFtpClient.cs:44]: And if an error will occur, stream won't be disposed too

File: HappyTravel.Komoro.JuniperChannelManager/Services/Updater/JuniperFtpClient.cs (hunk 5 above)
Comment (DFSko) [reply HappyTravel.Komoro.JuniperChannelManager/Services/Updater/JuniperFtpClient.cs]: added disconnecting

File: HappyTravel.Komoro.JuniperChannelManager/Services/Updater/JuniperFtpClient.cs (hunk 4 above)
Comment (DFSko) [reply HappyTravel.Komoro.JuniperChannelManager/Services/Updater/JuniperFtpClient.cs]: using compiles as try-finally, added using to ftp client

**Review by suhan3z — APPROVED** [review]

**Pull request merged**